# Gathering ENS Data


**Note:** This code uses part of my code publicly avaiable at: https://github.com/johnnatan-messias/arbitrum-analysis


In [1]:
import os
import json
import gzip
import pandas as pd
pd.set_option('display.precision', 18)
pd.set_option('max_colwidth', None)

In [2]:
data_dir = './data/'
os.makedirs(data_dir, exist_ok=True)

In [3]:
def load_addresses():
    with open(data_dir+'addresses.json', 'r') as file:
        top_addresses_dict = json.load(file)

    print("There are {} addresses in the dictionary".format(
        len(top_addresses_dict)))
    return top_addresses_dict

In [4]:
# Addresses already identified can be added to the json file ./data/addresses.json
addresses_dict = load_addresses()

There are 3191 addresses in the dictionary


## Gathering ENS data

For most projects, knowing the identity or ownership of public wallet addresses is crucial. Therefore, here I gathered information of ENS domains recorded on the Ethereum blockchain through an archive node.

From a total of 624,739 addresses in our Arbitrum data set, I was able to find 95,578 unique addresses that have an ENS domain registered on the Ethereum blockchain. These addresses are persisted in the ./data/addresses_dic.json.gz file.

Further, I collected a list of the ownership of 3191 addresses from the Etherscan API (top ETH holders), Compound website, and Sybil-List to compose our data of inferred addresses. These addresses are persisted in the json ./data/addresses.json file. However, we're not going to use them in this analysis but it shows that the code is ready for future projects. More details are available at section 4 of the following paper https://arxiv.org/pdf/2305.17655.pdf


In [5]:
from web3 import Web3
from ethereum import *

eth_node = os.environ['ETHEREUM_NODE_URL']
w3 = Web3(Web3.HTTPProvider(eth_node))
print("The most recent block number in the node's chain is", w3.eth.block_number)

The most recent block number in the node's chain is 19234234


In [6]:
ens_contract_address = "0x3671ae578e63fdf66ad4f3e12cc0c0d71ac7510c"

In [7]:
ens_contract = get_contract(
    w3, contract_address=Web3.to_checksum_address(ens_contract_address))

In [8]:
addresses_to_check = list(map(Web3.to_checksum_address, [
    '0x12bEc274578060E2E2701f1ec7DC3b7576e50cAC', '0x1f9090aaE28b8a3dCeaDf281B0F12828e676c326']))
addresses_to_check

['0x12bEc274578060E2E2701f1ec7DC3b7576e50cAC',
 '0x1f9090aaE28b8a3dCeaDf281B0F12828e676c326']

In [9]:
print("Getting ENS names for the addresses: ", addresses_to_check)
print(ens_contract.functions.getNames(addresses_to_check).call())

Getting ENS names for the addresses:  ['0x12bEc274578060E2E2701f1ec7DC3b7576e50cAC', '0x1f9090aaE28b8a3dCeaDf281B0F12828e676c326']
['rubesy999.eth', 'rsync-builder.eth']


In [10]:
print("Getting the ENS address for the ENS name rubesy999.eth")
w3.ens.address('rubesy999.eth').lower()

Getting the ENS address for the ENS name rubesy999.eth


'0x12bec274578060e2e2701f1ec7dc3b7576e50cac'

In [11]:
w3.ens.address('rubesy999.eth').lower() == addresses_to_check[0].lower()

True